### PR TITLE
VirtualHostOptions plugin

### PR DESCRIPTION
**Pull request by jbohanon**
# Description

Please include a high level summary of the changes.

This bug fixes ... \ This new feature can be used to ...

_Fill out any of the following sections that are relevant and remove the others_
## API changes
- Added x field to y resource
- ...

## Code changes
- Fix error in `Foo()` function
- Add `Bar()` function
- ...

## CI changes
- Adjusted schedule for x job
- ...

## Docs changes
- Added guide about feature x to public docs
- Updated README to account for y behavior
- ...

# Context

Users ran into this bug doing ... \ Users needed this feature to ...

See slack conversation [here](https://solo-io-corp.slack.com/archives/some/post)

## Interesting decisions
 
We chose to do things this way because ...

## Testing steps

I manually verified behavior by ...

## Notes for reviewers

Be sure to verify intended behavior by ...

Please proofread comments on ...

This is a complex PR and may require a huddle to discuss ...

# Checklist:

- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works

### CODE REVIEW

File: projects/gateway/pkg/translator/converter.go [hunk 1]
```diff
@@ -157,6 +157,7 @@ func (r *reporterHelper) addWarning(resource resources.InputResource, err error)
 // Since virtual services and route tables are often owned by different teams, it breaks multitenancy if
 // this function cannot return successfully; thus ALL ERRORS are added to the resource reports
 func (rv *routeVisitor) ConvertVirtualService(virtualService *gatewayv1.VirtualService, gateway *gatewayv1.Gateway, proxyName string, snapshot *gloov1snap.ApiSnapshot, reports reporter.ResourceReports) []*gloov1.Route {
+	panic("converting virtualservice")
```
Comment (jbohanon) [inline projects/gateway/pkg/translator/converter.go:160]: oops remove this
```suggestion
```

File: projects/gateway2/translator/plugins/virtualhostoptions/query/query.go [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+func (r *virtualHostOptionQueries) GetVirtualHostOptionsForGateway(ctx context.Context, gw *gwv1.Gateway, list *solokubev1.VirtualHostOptionList) error {
+	nn := types.NamespacedName{
+		Namespace: gw.Namespace,
+		Name:      gw.Name,
+	}
+	return r.c.List(
+		ctx,
+		list,
+		client.MatchingFieldsSelector{Selector: fields.OneTermEqualSelector(VirtualHostOptionTargetField, nn.String())},
+		client.InNamespace(gw.GetNamespace()),
```
[the remark this answers is not in the document]
Comment (jbohanon) [reply projects/gateway2/translator/plugins/virtualhostoptions/query/query.go:37]: I am not sure why this restriction is imposed. It exists on the RouteOptions queries, and it was brought over as it was there. If this is not needed on VirtualHostOptions queries, I can remove it and look into reference grants.

@lgadban do you have more context on the namespace restriction?

File: projects/gateway2/translator/plugins/virtualhostoptions/query/indexers.go [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+	targetRef := rtOpt.Spec.GetTargetRef()
+	if targetRef == nil {
+		return res
+	}
+	if targetRef.GetGroup() != gwv1.GroupName {
+		return res
+	}
+	if targetRef.GetKind() != wellknown.HTTPRouteKind {
+		return res
+	}
```
[the remark this answers is not in the document]
Comment (jbohanon) [reply projects/gateway2/translator/plugins/virtualhostoptions/query/indexers.go:37]: Good catch; will fix

File: projects/gateway2/translator/plugins/virtualhostoptions/query/query.go [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+	"k8s.io/apimachinery/pkg/types"
+
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	gwv1 "sigs.k8s.io/gateway-api/apis/v1"
+)
+
+type VirtualHostOptionQueries interface {
+	// Populates the provided VirtualHostOptionList with the VirtualHostOption resources attached to the provided Gateway.
+	// Note that currently, only VirtualHostOptions in the same namespace as the Gateway can be attached.
+	GetVirtualHostOptionsForGateway(ctx context.Context, gw *gwv1.Gateway, list *solokubev1.VirtualHostOptionList) error
```
[the remark this answers is not in the document]
Comment (jbohanon) [reply projects/gateway2/translator/plugins/virtualhostoptions/query/query.go:17]: It looks like it was done this way because the controller-runtime client's `List` accepts the list pointer as an output parameter. I agree it's not needed so I will restructure this to return the list

File: projects/gateway2/translator/plugins/virtualhostoptions/query/query.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (lgadban) [reply projects/gateway2/translator/plugins/virtualhostoptions/query/query.go:37]: Enforcing same namespace was the simplest possible solution (without requiring reference grants) and as we are not merging Route/VHost options, there can only be one policy object attached, so it seems reasonable.

We can add reference grants for cross-namespace if we need to, but i think it is lower priority than e.g. adding status reporting for VHostOption resources

File: projects/gateway2/translator/plugins/virtualhostoptions/query/query.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (lgadban) [reply projects/gateway2/translator/plugins/virtualhostoptions/query/query.go:17]: Yes this was done to match the `List` semantics. If we change to instantiate the List inside the query, can we also update the RouteOption query to do the same? Would prefer them to be consistent

File: projects/gateway2/translator/plugins/virtualhostoptions/virtualhost_options_plugin.go [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+
+	var optToUse *solokubev1.VirtualHostOption
+	// If there is not a section name or the specified section name matches our listener, apply the vhost options
+	if targetedOpt, ok := optsWithSectionName[string(listenerCtx.GwListener.Name)]; ok {
+		optToUse = targetedOpt
+	} else if len(optsWithoutSectionName) == 1 {
+		optToUse = optsWithoutSectionName[0]
+	}
+	var vhs []*v1.VirtualHost
+	switch outListener.GetListenerType().(type) {
```
Comment (lgadban) [inline projects/gateway2/translator/plugins/virtualhostoptions/virtualhost_options_plugin.go:73]: Possible nit: I'm fairly certain we are only rendering out aggregateListeners; we may want to codify that and error if we encounter any of other type. The defensiveness here is useful but might just be more confusing to build in support for types that may never be used

File: projects/gateway2/translator/plugins/virtualhostoptions/virtualhost_options_suite_test.go [hunk 6]
```diff
@@ -0,0 +1,13 @@
...
+import (
+	"testing"
+
+	. "github.com/onsi/ginkgo/v2"
+	. "github.com/onsi/gomega"
+)
+
+func TestRouteOptionsPlugin(t *testing.T) {
+	RegisterFailHandler(Fail)
+	RunSpecs(t, "RouteOptions Plugin Suite")
```
Comment (lgadban) [inline projects/gateway2/translator/plugins/virtualhostoptions/virtualhost_options_suite_test.go:12]: leftover RouteOptions

File: projects/gateway2/translator/plugins/virtualhostoptions/virtualhost_options_plugin.go [hunk 7]
```diff
@@ -0,0 +1,128 @@
...
+	ctx context.Context,
+	listenerCtx *plugins.ListenerContext,
+	outListener *v1.Listener,
+) error {
+
+	// attachedOptions represents all VirtualHostOptions targeting the Gateway on which this listener resides
+	attachedOptions := getAttachedVirtualHostOptions(ctx, listenerCtx.Gateway, p.vhOptQueries)
+	if len(attachedOptions) == 0 {
+		return nil
+	}
```
[the remark this answers is not in the document]
Comment (jbohanon) [reply projects/gateway2/translator/plugins/virtualhostoptions/virtualhost_options_plugin.go]: Since we are selecting for a single options here, by adding this selection logic into the query it no longer makes sense for the query to return a List. @lgadban does it still make sense to have the queries match format between RouteOptions and VirtualHostOptions given this deviation, or have I misinterpreted the comment here?

File: projects/gateway2/translator/plugins/virtualhostoptions/virtualhost_options_plugin.go (hunk 5 above)
Comment (jbohanon) [reply projects/gateway2/translator/plugins/virtualhostoptions/virtualhost_options_plugin.go]: Good call, added an error for anything but an aggregate listener

TODO: test that error

File: projects/gateway2/translator/plugins/virtualhostoptions/virtualhost_options_suite_test.go (hunk 6 above)
Comment (jbohanon) [reply projects/gateway2/translator/plugins/virtualhostoptions/virtualhost_options_suite_test.go]: Thanks; removed

File: projects/gateway2/translator/plugins/virtualhostoptions/virtualhost_options_plugin.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (EItanya) [reply projects/gateway2/translator/plugins/virtualhostoptions/virtualhost_options_plugin.go]: My question here is slightly different I think. I'm not arguing that we should return a single `opt` here, but rather all matching opts considering the entirety of the matching logic. The plugin can decide how to update statuses etc, but I don't think half of the selection logic should be done by the query engine, and the other half by the plugin.

File: projects/gateway2/translator/plugins/virtualhostoptions/query/query_test.go [hunk 8]
```diff
@@ -0,0 +1,198 @@
...
+			TargetRef: &corev1.PolicyTargetReferenceWithSectionName{
+				Group:     gwv1.GroupVersion.Group,
+				Kind:      wellknown.GatewayKind,
+				Name:      "test",
+				Namespace: wrapperspb.String("default"),
+			},
+			Options: &v1.VirtualHostOptions{},
+		},
+	}
+}
```
Comment (lgadban) [inline projects/gateway2/translator/plugins/virtualhostoptions/query/query_test.go:140]: should we add a test for sectionName being present?

File: projects/gateway2/translator/plugins/virtualhostoptions/virtualhost_options_plugin_test.go [hunk 9]
```diff
@@ -0,0 +1,211 @@
...
+			Options: &v1.VirtualHostOptions{
+				Retries: &retries.RetryPolicy{
+					RetryOn:    "5xx",
+					NumRetries: 5,
+				},
+			},
+		},
+	}
+}
+
```
Comment (lgadban) [inline projects/gateway2/translator/plugins/virtualhostoptions/virtualhost_options_plugin_test.go:167]: should we add a test for sectionName being present?

**Review by lgadban — COMMENTED** [review]
mostly LGTM, a few comments/questions

File: projects/gateway2/translator/plugins/virtualhostoptions/virtualhost_options_plugin.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (lgadban) [reply projects/gateway2/translator/plugins/virtualhostoptions/virtualhost_options_plugin.go]: I think I had the same concern but I believe it has been addressed in a more recent commit.

Current state is the field used for selection during Listing is targetRef but only the NamespacedName of the `Gateway`. Then the sectionName processing is done within the query, post-cache List.

I suppose there is theoretically another layer of indexing which could be used List attached policy by sectionName, but that would require Listing per Listener.

**Comment by lgadban** [comment]
Worth calling out a couple of things:
* the current implementation is a Listener plugin, but it's really in the context of a gloov1.Listener, not a gwapi.Listener

Separately, there is a (somewhat) similar naivety in this implementation (similar to the current state of RouteOptions) where:
* iterate for every gloov1.Listener
* run a new query for VHostOpts attached to a parent gwapi.Gateway
* results in a new DeepCopy of the gwapi.Gateway, regardless of how many gloov1.Listeners are from the same parent gwapi.Gateway

**Comment by EItanya** [comment]
> results in a new DeepCopy of the gwapi.Gateway, regardless of how many gloov1.Listeners are from the same parent gwapi.Gateway

I don't believe this is a problem with this PR specifically, but rather that DeepCopy needs to be disabled for the project which is a work item in itself

File: projects/gateway2/translator/plugins/virtualhostoptions/virtualhost_options_plugin.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (EItanya) [reply projects/gateway2/translator/plugins/virtualhostoptions/virtualhost_options_plugin.go]: Yea I think the indexing is fine for now, especially given that needing to index by sectionName could end up taking up more memory than it gives value. I added another comment about returning all selected options from the query though because I don't think it's the responsibility of the query to decide how many options is the right number.

[the remark this answers is not in the document]
Comment (jbohanon) [reply projects/gateway2/translator/plugins/virtualhostoptions/virtualhost_options_plugin.go]: Agreed. I made modifications to the query to keep the selection logic but defer the discernment to the plugin. Is this a more appropriate separation of concerns?

File: projects/gateway2/translator/listener/gateway_listener_translator.go [hunk 10]
```diff
@@ -178,6 +182,16 @@ func (ml *mergedListeners) translateListeners(
 	var listeners []*v1.Listener
 	for _, mergedListener := range ml.listeners {
 		listener := mergedListener.translateListener(ctx, pluginRegistry, queries, reporter)
+
+		// run listener plugins
+		for _, listenerPlugin := range pluginRegistry.GetListenerPlugins() {
+			listenerPlugin.ApplyListenerPlugin(ctx, &plugins.ListenerContext{
```
[the remark this answers is not in the document]
Comment (jbohanon) [reply projects/gateway2/translator/listener/gateway_listener_translator.go]: Added!

File: projects/gateway2/translator/plugins/plugins.go [hunk 11]
```diff
@@ -39,6 +39,25 @@ type RoutePlugin interface {
 	) error
 }
 
+type ListenerContext struct {
+	// top-level Gateway
+	Gateway *gwv1.Gateway
+	// gw Listener being processed
+	GwListener *gwv1.Listener
+	// Reporter for the correct ParentRef associated with this Listener
+	Reporter reports.Reporter
```
[the remark this answers is not in the document]
Comment (jbohanon) [reply projects/gateway2/translator/plugins/plugins.go:48]: Good call; I removed the item from the context struct.

File: projects/gateway2/translator/plugins/virtualhostoptions/query/query_test.go (hunk 8 above)
Comment (jbohanon) [reply projects/gateway2/translator/plugins/virtualhostoptions/query/query_test.go:213]: I added some tests, lmk what you think of the coverage.

File: projects/gateway2/translator/plugins/virtualhostoptions/virtualhost_options_plugin_test.go (hunk 9 above)
Comment (jbohanon) [reply projects/gateway2/translator/plugins/virtualhostoptions/virtualhost_options_plugin_test.go:188]: I added a test here, but I felt like the query tests covered a lot of the corner case, so equal coverage here would be largely redundant. wdyt?

**Comment by jbohanon** [comment]
> * run a new query for VHostOpts attached to a parent gwapi.Gateway

Is it worth caching the List result from a query on the query object keyed on the gateway NN? I'm not sure how the `controller-runtime` query engine works or if that would be an over-extension of the query implementation

File: projects/gateway2/translator/plugins/virtualhostoptions/query/query.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (jbohanon) [reply projects/gateway2/translator/plugins/virtualhostoptions/query/query.go]: If this merges as-is I will follow up with another PR to make the RouteOptions consistent

File: projects/gateway2/translator/listener/gateway_listener_translator.go [hunk 12]
```diff
@@ -185,11 +186,14 @@ func (ml *mergedListeners) translateListeners(
...
 		// run listener plugins
 		for _, listenerPlugin := range pluginRegistry.GetListenerPlugins() {
-			listenerPlugin.ApplyListenerPlugin(ctx, &plugins.ListenerContext{
+			err := listenerPlugin.ApplyListenerPlugin(ctx, &plugins.ListenerContext{
 				Gateway:    &mergedListener.parentGateway,
 				GwListener: &mergedListener.listener,
 				Reporter:   reporter,
 			}, listener)
+			if err != nil {
+				contextutils.LoggerFrom(ctx).Errorf("error in ListenerPlugin: %v", err)
```
Comment (EItanya) [inline projects/gateway2/translator/listener/gateway_listener_translator.go:195]: If these start erroring continuously this could get very noisy, is there any kind of reporting we want to do with this?

File: projects/gateway2/translator/plugins/virtualhostoptions/query/query.go [hunk 13]
```diff
@@ -16,21 +15,26 @@ type VirtualHostOptionQueries interface {
...
 }
 
 type virtualHostOptionQueries struct {
 	c client.Client
 }
 
+type VirtualHostOptionsQueryResult struct {
+	OptsWithSectionName    []*solokubev1.VirtualHostOption
+	OptsWithoutSectionName []*solokubev1.VirtualHostOption
+}
```
Comment (EItanya) [inline projects/gateway2/translator/plugins/virtualhostoptions/query/query.go:28]: Ok I'm confused, why are we distinguishing between these 2?

Comment (jbohanon) [reply projects/gateway2/translator/plugins/virtualhostoptions/query/query.go:28]: An option with a section name should always override an option without a section name, but that seemed to fall under discernment re: prior conversation

File: projects/gateway2/translator/listener/gateway_listener_translator.go (hunk 12 above)
Comment (jbohanon) [reply projects/gateway2/translator/listener/gateway_listener_translator.go:195]: It's a good point. @lgadban is this something that upcoming status reporting would handle? I'm guessing not since the current implementation of the RouteOptions still prints this error after status reporting was added? Or was that an oversight?

File: projects/gateway2/translator/plugins/virtualhostoptions/query/query.go (hunk 13 above)
Comment (EItanya) [reply projects/gateway2/translator/plugins/virtualhostoptions/query/query.go:28]: Ahhhh I see the problem, very interesting. I may have sent you on a bit of a goose chase. What if we return them in order of specificity? That way that logic to determine can be `libraried` but it's also clear which one the caller should favor? @lgadban what do you think of that?

Comment (jbohanon) [reply projects/gateway2/translator/plugins/virtualhostoptions/query/query.go:28]: I think that could still make it difficult to determine invalid config states such as multiple options targeting with section name, or multiple options targeting the gateway but without section name. 
IMO the four valid config states are:
- No options at all
- one option attached to gateway without section name, none with section name
- one option attached to gateway with section name, none without section name
- two options attached to gateway, one without section name, one with section name

File: projects/gateway2/translator/listener/gateway_listener_translator.go (hunk 12 above)
Comment (lgadban) [reply projects/gateway2/translator/listener/gateway_listener_translator.go:194]: The intention was to log in addition to reporting as (IMO at least) they are used for different things. Probably worth a larger discussion for the holistic approach, that being said... as we don't have VHostOption status persistence, I think it makes even more sense to log in this specific case.

File: projects/gateway2/translator/plugins/virtualhostoptions/query/query.go (hunk 13 above)
Comment (lgadban) [reply projects/gateway2/translator/plugins/virtualhostoptions/query/query.go:28]: I am okay with the current logic as this query is specifically receiving a Listener **and a Gateway**. There should only ever be a single `VirtualHost`

If we do not want to do the sectionName/Listener logic in the query, I would say let's change the query to be `getAttachedVHostToGateway` and return all VHostOptions attached to a Gateway. This also gives us the option to report conflicted status (eventually) on the VHostOptions that are not being attached (as we should only process the earliest possible VHost)

Comment (EItanya) [reply projects/gateway2/translator/plugins/virtualhostoptions/query/query.go:28]: Why is section name treated differently, isn't it just a more specific way of targeting?

Comment (lgadban) [reply projects/gateway2/translator/plugins/virtualhostoptions/query/query.go:28]: Yeah, nevermind, I didn't see the change where we are returning the two lists.

If we keep the query as-is, we should only return a single VirtualHostOption, the most-specific. Is this what you were initially suggesting @EItanya ?

This does mean that eventually, if there are more than one at the level of specificity, we will need to e.g. pass a reporter to report on the VHostOpts that were targeted but weren't used because an earlier object existed etc

Comment (jbohanon) [reply projects/gateway2/translator/plugins/virtualhostoptions/query/query.go:28]: I guess to @EItanya's point if they are sorted by specificity (not sure what is more specific if two target the same gateway with same section name), then it makes sense to just return the sorted list, use the head of it, and report on all of the others

Comment (lgadban) [reply projects/gateway2/translator/plugins/virtualhostoptions/query/query.go:28]: that's fair but I struggle to see value of doing it that way. It would only be to not pass in a report mechanism (but we do pass that in other gateway queries, notably)

Comment (jbohanon) [reply projects/gateway2/translator/plugins/virtualhostoptions/query/query.go:28]: I guess the question is whose responsibility is it to determine what constitutes conflicting vhost options attachments, the plugin or the query

Comment (jbohanon) [reply projects/gateway2/translator/plugins/virtualhostoptions/query/query.go:28]: The decision made at the EOD meeting yesterday was to run with the suggestion of returning a sorted list in order to unblock this initial implementation. We can change this implementation detail later as needed

File: projects/gateway2/translator/listener/gateway_listener_translator.go (hunk 12 above)
Comment (jbohanon) [reply projects/gateway2/translator/listener/gateway_listener_translator.go:194]: We will keep this logging in AT LEAST until status reporting is implemented, so it will stay in for this PR.

**Comment by jbohanon** [comment]
> > * run a new query for VHostOpts attached to a parent gwapi.Gateway
> 
> Is it worth caching the List result from a query on the query object keyed on the gateway NN? I'm not sure how the `controller-runtime` query engine works or if that would be an over-extension of the query implementation

This is outside the scope of this PR

File: projects/gateway2/translator/plugins/virtualhostoptions/query/query.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (jbohanon) [reply projects/gateway2/translator/plugins/virtualhostoptions/query/query.go]: Support for reference grants can be assessed and added later as needed

File: projects/gateway2/translator/plugins/plugins.go [hunk 14]
```diff
@@ -39,6 +39,23 @@ type RoutePlugin interface {
...
+	// gw Listener being processed
+	GwListener *gwv1.Listener
+}
+type ListenerPlugin interface {
+	Plugin
+
+	// ApplyListenerPlugin is called for each Listener in a Gateway
+	ApplyListenerPlugin(
+		ctx context.Context,
+		routeCtx *ListenerContext,
```
Comment (ilackarms) [inline projects/gateway2/translator/plugins/plugins.go:54]: `listenerCtx`

File: projects/gateway2/translator/plugins/virtualhostoptions/virtualhost_options_plugin.go [hunk 15]
```diff
@@ -0,0 +1,73 @@
...
+var (
+	ErrUnexpectedListenerType = eris.New("unexpected listener type")
+	errUnexpectedListenerType = func(l *v1.Listener) error {
+		return eris.Wrapf(ErrUnexpectedListenerType, "expected AggregateListener, got %T", l.GetListenerType())
+	}
+)
+
+func NewPlugin(gwQueries gwquery.GatewayQueries, client client.Client) *plugin {
+	return &plugin{
+		gwQueries,
```
Comment (ilackarms) [inline projects/gateway2/translator/plugins/virtualhostoptions/virtualhost_options_plugin.go:31]: nit: prefer to always use field names when constructing structs

**Review by EItanya — APPROVED** [review]

**Review by lgadban — APPROVED** [review]

**Comment by jbohanon** [comment]
Paging Dr. Dozer
